apps/shell/app.js

```diff
@@ -78,8 +78,13 @@ default_sshhost = process.env.DEFAULT_SSHHOST || default_sshhost;
 if (default_sshhost) host_whitelist.add(default_sshhost);
 function host_and_dir_from_url(url){
   let match = url.match(host_path_rx),
-  hostname = match[1] === "default" ? default_sshhost : match[1],
-  directory = match[2] ? decodeURIComponent(match[2]) : null;
+  hostname = default_sshhost, 
+  directory = null;
+
+  if (match) {
+    hostname = match[1] === "default" ? default_sshhost : match[1];
+    directory = match[2] ? decodeURIComponent(match[2]) : null;
+  }
```

Comment:
So if the URL doesn't match for any reason, we use default host and directory null. I don't know why that makes me nervous. Seems to broad. I don't have a better suggestion at this point. @johrstrom any suggestion?

Reply:
The only way this fails is if there's no default host set in `/etc/ood/config/clusters.d/*.y*ml` We could add additional logic that goes if no default host is set in yaml configs or through an env var then default to use first host in y*ml configs

Reply:
I like the first host idea. So we would do

   default_sshhost = process.env.DEFAULT_SSHHOST || default_sshhost || first_available_host;

Reply:
The host allowlist is a Set so there is no 'first' host. We can do the first host found when going in the cluster configs although problems may arise when the host is unavailable for general access even if there is a present cluster config (ruby.osc.edu in our case). This is a very specific edge case.
However this shouldn't be a problem if we make the `default: ` field necessary for cluster configs.

Reply:
I like the implementation you added (except for one minor improvement suggested above). The first available host is useful because that will be the case every new site experiences when installing OnDemand and adding their first cluster config. A default option in the login section of the cluster config is a new concept, so no sites will have this set when installing OnDemand, so it shouldn't be required. As for "the host is unavailable for general access" I think we will address that with 2.0 - getting rid of those atrocious Ruby enforced ACLs and asking people to use file permissions on the cluster configs like we do on the apps.